modules/dnn/src/layers/nms_layer.cpp

```diff
@@ -0,0 +1,154 @@
+
+        int _num_batches = inputs[0][0];
+        int _num_classes = inputs[1][1];
+        // outputs[0]: selected_indices, num_selected_indices * [batch_index, class_index, box_index]
+        // consider the case whose _num_batches == 1 & _num_classes == 1
+        outputs.resize(1, shape(keep_top_k, 3));
+
+        return false;
+    }
+
+    static inline float rect2dOverlap(const Rect2d& a, const Rect2d& b)
+    {
+        return 1.f - static_cast<float>(jaccardDistance(a, b));
+    }
+
+    void forward(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr, OutputArrayOfArrays internals_arr) CV_OVERRIDE
+    {
+        CV_TRACE_FUNCTION();
+        CV_TRACE_ARG_VALUE(name, "name", name.c_str());
+
+        std::vector<Mat> inputs, outputs;
+        inputs_arr.getMatVector(inputs);
+        outputs_arr.getMatVector(outputs);
+
+        int num_batches = inputs[0].size[0];
+        int num_boxes = inputs[0].size[1];
+
+        std::vector<Rect2d> boxes;
+        std::vector<float> scores;
+        // Retrieve bboxes
+        boxes.resize(num_boxes);
+        const float* ptr_boxes = (float*)inputs[0].data;
+        if (center_point_box == 1) // num_boxes * [cx, cy, w, h]
+        {
+            float cx, cy, w, h;
+            for (size_t i = 0; i < boxes.size(); i++)
+            {
+                Rect2d& box = boxes[i];
+                cx = ptr_boxes[i * 4];
+                cy = ptr_boxes[i * 4 + 1];
+                w  = ptr_boxes[i * 4 + 2];
+                h  = ptr_boxes[i * 4 + 3];
+
+                box.x = cx - 0.5 * w;
+                box.y = cy - 0.5 * h;
+                box.width = w;
+                box.height = h;
+            }
+        }
+        else // num_boxes * [y1, x1, y2, x2]
+        {
+            float x1, y1, x2, y2;
+            for (size_t i = 0; i < boxes.size(); i++)
+            {
+                Rect2d& box = boxes[i];
+                y1 = ptr_boxes[i * 4];
+                x1 = ptr_boxes[i * 4 + 1];
+                y2 = ptr_boxes[i * 4 + 2];
+                x2 = ptr_boxes[i * 4 + 3];
+
+                box.x = x1;
+                box.y = y1;
+                box.width = x2 - x1;
+                box.height = y2 - y1;
+            }
+        }
+        // Retrieve scores
+        const float* ptr_scores = (float*)inputs[1].data;
+        if (inputs[1].isContinuous())
+        {
+            std::cout << "It is continuous!!!" << std::endl;
+            scores.assign(ptr_scores, ptr_scores + inputs[1].total());
+        }
+        else
+        {
+            scores.resize(num_boxes);
+            for (size_t i = 0; i < scores.size(); i++)
+            {
+                scores[i] = ptr_scores[i];
+            }
+        }
+
+        // NMS
+        std::vector<int> keep_indices;
+        NMSFast_(boxes, scores, score_threshold, iou_threshold, 1.0, top_k, keep_indices, rect2dOverlap, keep_top_k);
+
+        // Store to output
+        outputs[0].setTo(-1);
```

Comment:
Since dnn does not support dynamic shape for now, the number of output boxes have to be set fixed (650 is used in this implementation, but there will not be so many boxes in one single image typically speaking). So invalid indices have to be set to -1.

@rogday if the nms operator is ported from `torchvision.ops.nms`, the NonMaxSuppression operator is always followed by a Gather operator (see the visualization in https://github.com/opencv/opencv_extra/pull/1005). I wonder if it is reasonable to drop those invalid indices in your pull request for the Gather operator.